updater/fetchers/alpine/alpine.go

```diff
@@ -0,0 +1,247 @@
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/coreos/pkg/capnslog"
+
+	"github.com/coreos/clair/database"
+	"github.com/coreos/clair/updater"
+	"github.com/coreos/clair/utils"
+	cerrors "github.com/coreos/clair/utils/errors"
+	"github.com/coreos/clair/utils/types"
+)
+
+const (
+	secdbGitURL  = "http://git.alpinelinux.org/cgit/alpine-secdb"
+	updaterFlag  = "alpine-secdbUpdater"
+	nvdURLPrefix = "https://cve.mitre.org/cgi-bin/cvename.cgi?name="
+)
+
+var (
+	// ErrFilesystem is returned when a fetcher fails to interact with the local filesystem.
+	ErrFilesystem = errors.New("updater/fetchers: something went wrong when interacting with the fs")
+
+	// ErrGitFailure is returned when a fetcher fails to interact with git.
+	ErrGitFailure = errors.New("updater/fetchers: something went wrong when interacting with git")
+
+	log = capnslog.NewPackageLogger("github.com/coreos/clair", "updater/fetchers/alpine")
+)
+
+func init() {
+	updater.RegisterFetcher("alpine", &fetcher{})
+}
+
+type fetcher struct {
+	repositoryLocalPath string
+}
+
+func (f *fetcher) FetchUpdate(db database.Datastore) (resp updater.FetcherResponse, err error) {
+	log.Info("fetching Alpine vulnerabilities")
+
+	// Pull the master branch.
+	var commit string
+	commit, err = f.pullRepository()
+	if err != nil {
+		return
+	}
+
+	// Ask the database for the latest commit we successfully applied.
+	var dbCommit string
+	dbCommit, err = db.GetKeyValue(updaterFlag)
+	if err != nil {
+		return
+	}
+
+	// Set the updaterFlag to equal the commit processed.
+	resp.FlagName = updaterFlag
+	resp.FlagValue = commit
+
+	// Short-circuit if there have been no updates.
+	if commit == dbCommit {
+		log.Debug("no alpine update")
+		return
+	}
+
+	// Append any changed vulnerabilities to the response.
+	for _, namespace := range []string{"v3.3", "v3.4"} {
+		var file io.ReadCloser
+		file, err = os.Open(f.repositoryLocalPath + "/" + namespace + "/main.yaml")
+		if err != nil {
+			return
+		}
+		log.Debug(namespace)
+
+		var vulns []database.Vulnerability
+		switch namespace {
+		case "v3.3":
+			vulns, err = parse33YAML(file)
+		case "v3.4":
+			vulns, err = parse34YAML(file)
+		}
+		if err != nil {
+			return
+		}
+
+		resp.Vulnerabilities = append(resp.Vulnerabilities, vulns...)
+		file.Close()
+	}
+
+	return
+}
+
+func (f *fetcher) pullRepository() (commit string, err error) {
+	if _, pathExists := os.Stat(f.repositoryLocalPath); f.repositoryLocalPath == "" || os.IsNotExist(pathExists) {
+		if f.repositoryLocalPath, err = ioutil.TempDir(os.TempDir(), "alpine-secdb"); err != nil {
+			return "", ErrFilesystem
+		}
+
+		if out, err := utils.Exec(f.repositoryLocalPath, "git", "clone", secdbGitURL, "."); err != nil {
+			f.Clean()
+			log.Errorf("could not pull alpine-secdb repository: %s. output: %s", err, out)
+			return "", cerrors.ErrCouldNotDownload
+		}
+	}
+
+	out, err := utils.Exec(f.repositoryLocalPath, "git", "rev-parse", "HEAD")
```

Comment:
So the first time, we `clone`. The next times, we just `rev-parse HEAD`. Shouldn't we pull before? Otherwise even if `HEAD` moved, we'd still be reading the old files when opening them.